syntax = "proto3";

package algotrader.model2;


message DoubleTimeSeries{
    string series_id = 1;
    string frame_id = 2;
    repeated int64 timestamp = 10;
    repeated double values = 11;
}

message Int32TimeSeries{
    string series_id = 1;
    string frame_id = 2;
    repeated int64 timestamp = 10;
    repeated int32 values = 11;
}

message Int64TimeSeries{
    string series_id = 1;
    string frame_id = 2;
    repeated int64 timestamp = 10;
    repeated int64 values = 11;
}

message BoolTimeSeries{
    string series_id = 1;
    string frame_id = 2;
    repeated int64 timestamp = 10;
    repeated bool values = 11;
}

message StringTimeSeries{
    string series_id = 1;
    string frame_id = 2;
    repeated int64 timestamp = 10;
    repeated string values = 11;
}


message DataFrame{
    string id = 1;
    repeated int64 index = 2;
     map<string, DoubleTimeSeries> double_data = 10;
     map<string, Int32TimeSeries> int_data = 11;
     map<string, Int64TimeSeries> long_data = 12;
     map<string, BoolTimeSeries> bool_data = 13;
     map<string, StringTimeSeries> string_data = 14;


}

message TimeSeriesUpdateEvent{
    string source = 1;
    TimeSeriesItem item= 2;
}

message TimeSeries{
    message Input{
        string source = 1;
        repeated string keys = 2;
    }

    string series_id = 1;
    string series_cls = 2;
    repeated string keys = 3;
    string desc = 4;
    repeated Input inputs = 5;
    string default_output_key = 6;
    double missing_value_replace = 7;
    int64 start_time = 8;
    int64 end_time = 9;
    repeated TimeSeriesItem items = 10;
     map<string, string> configs = 11;
}

